lib/doctest.js

```diff
@@ -488,21 +504,31 @@ function functionEval(source) {
   var evaluate = Function ('__doctest', source);
   var queue = [];
   evaluate ({enqueue: function(io) { queue.push (io); }});
-  return run (queue);
+  return Promise.resolve (run (queue));
 }
 
 function commonjsEval(source, path) {
   var abspath =
   (pathlib.resolve (path)).replace (/[.][^.]+$/, '-' + Date.now () + '.js');
 
-  fs.writeFileSync (abspath, source);
-  var queue;
-  try {
-    queue = (require (abspath)).__doctest.queue;
-  } finally {
-    fs.unlinkSync (abspath);
-  }
-  return run (queue);
+  return util.promisify (fs.writeFile) (abspath, source).then(function() {
+    return run (require (abspath)).__doctest.queue;
+  }).finally (function() {
+    return util.promisify (fs.unlink) (abspath);
+  });
+}
+
+function esmEval(source, path) {
+  var abspath =
+  (pathlib.resolve (path)).replace (/[.][^.]+$/, '-' + Date.now () + '.mjs');
+
+  return util.promisify (fs.writeFile) (abspath, source).then(function() {
+    return import (abspath);
```

Comment:
Here's where the magic was supposed to happen - but it turns out it's not that easy. I had assumed that the dynamic `import` function would be available to commonjs scripts in Node. But it's not. In order to achieve what I'm trying to achieve here, this whole file would have to be rewritten to `.mjs`.

I think there might be a way to do it without too much effort, keeping the commonjs and amd support in a `.js` file which is loaded by an overarching `.mjs` file which adds esm support. I'll look into that sometime.

Reply:
This seems to be the deepest change, i'm not sure that i can help with analysis here, i guess that we would like to compare the costs of changing with the demand for the feature, @davidchambers can provide a better scope :)